src/python/platforms/android/emulator.py

```diff
@@ -0,0 +1,126 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Helper functions for running commands on emulated Android devices."""
+
+import os
+import shutil
+import subprocess
+import tempfile
+import time
+
+from google_cloud_utils import storage
+from metrics import logs
+from platforms.android import adb
+from system import archive
+from system import environment
+from system import new_process
+
+_WAIT_SECONDS = 10
+
+_emu_proc = None
+_emu_users = 0
+
+
+class EmulatorError(Exception):
+  """Error for errors handling the Android emulator."""
+
+
+class EmulatorProcess(object):
+  """A EmulatorProcess encapsulates the creation, running, and destruction
+  of Android emulator processes."""
+
+  def __init__(self):
+    self.process_runner = None
+    self.popen = None
+    self.logfile = None
+
+    log_path = os.path.join(tempfile.gettempdir(), 'android-emulator-log')
+    self.logfile = open(log_path, 'wb')
+
+  def create(self):
+    """Configures a emulator process which can subsequently be `run`."""
+    # Download emulator image
+    temp_directory = environment.get_value('BOT_TMPDIR')
+
+    archive_src_path = environment.get_value('ANDROID_EMULATOR_BUCKET_PATH')
+    archive_dst_path = os.path.join(temp_directory, 'emulator_bundle.zip')
+
+    storage.copy_file_from(archive_src_path, archive_dst_path)
+
+    # Extract emulator image.
+    self.emu_path = os.path.join(temp_directory, 'emulator')
+    archive.unpack(archive_dst_path, self.emu_path)
+    os.remove(archive_dst_path)
+
+    # Run emulator
+    script_path = os.path.join(
+        self.emu_path, environment.get_value('ANDROID_EMULATOR_SCRIPT_PATH'))
+
+    self.process_runner = new_process.ProcessRunner(script_path, [])
+
+  def run(self):
+    """Actually runs a emulator, assuming `create` has already been called."""
+    if not self.process_runner:
+      raise EmulatorError('Attempted to `run` emulator before calling `create`')
+
+    devices_before = adb.get_devices()
+    new_device = False
+
+    logs.log('Starting emulator.')
+    self.popen = self.process_runner.run(
+        stdout=self.logfile, stderr=subprocess.PIPE)
+
+    logs.log('Waiting for emulated device to come online.')
+    while not new_device:
```

Comment:
Will we be starting one emulator at a time, this looks racy if another script is running and trying to start at same time. There should be some way to specify our own serial for emulator or prefix, then can just wait on it.

Reply:
The problem here is that we don't or can't know in advance what the serial of the emulated device will be, since it's determined when the emulator starts.

Reply:
Good point. I've modified the startup script to print out the device serial number, which is determined at runtime, and added code CF to read it.